pkg/storaged/client.js

```diff
@@ -206,6 +214,175 @@
+            // https://github.com/storaged-project/udisks/commit/b6966b7076cd837f9d307eef64beedf01bc863ae
+            try {
+                output = await cockpit.spawn(["btrfs", "subvolume", "get-default", mount_point], { superuser: true, err: "message" });
+                const id_match = output.match(/ID (\d+).*/);
+                if (id_match)
+                    btrfs_default_subvol[uuid] = Number(id_match[1]);
+            } catch (err) {
+                console.error(`unable to obtain default subvolume for mount point ${mount_point}`, err);
+            }
+
+            // HACK: UDisks should expose a better btrfs API with btrfs device information
+            // https://github.com/storaged-project/udisks/issues/1232
+            // TODO: optimise into just parsing one `btrfs filesystem show`?
+            const usages = {};
+            let usage_output;
+            try {
+                usage_output = await cockpit.spawn(["btrfs", "filesystem", "show", "--raw", uuid], { superuser: true, err: "message" });
+            } catch (err) {
+                console.error(`btrfs filesystem show ${uuid}`, err);
+                return;
+            }
+            for (const line of usage_output.split("\n")) {
+                const match = usage_regex.exec(line);
+                if (match) {
+                    const { used, device } = match.groups;
+                    usages[device] = used;
+                }
+            }
+            uuids_usage[uuid] = usages;
+        } else {
+            uuids_subvols[uuid] = null;
+            uuids_usage[uuid] = null;
+        }
+    }
+
+    // HACK: we should try to use --poll and run this separate from the 5 seconds btrfs subvolume poll
+    // This fails when no btrfs filesystem is found with the --mtab option and exits with 1, so that is kinda useless, however without --mtab
+    // we don't get a nice flat structure. So we ignore the errors
+    if (Object.keys(client.uuids_btrfs_volume).length > 0) {
+        let findmnt_output;
+        try {
+            findmnt_output = await cockpit.spawn(["findmnt", "--type", "btrfs", "--mtab", "-o", "UUID,OPTIONS,SOURCE,TARGET", "--json"],
+                                                 { superuser: true, err: "message" });
+            try {
+                // Extract the data into a { uuid: { subvolid: { subvol, target } } }
+                const mounts = JSON.parse(findmnt_output);
+                if ("filesystems" in mounts) {
+                    for (const fs of mounts.filesystems) {
+                        const subvolid_match = fs.options.match(/subvolid=(?<subvolid>\d+)/);
+                        const subvol_match = fs.options.match(/subvol=(?<subvol>[\w\\/]+)/);
+                        if (!subvolid_match && !subvol_match) {
+                            console.warn("findmnt entry without subvol and subvolid", fs);
+                            break;
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19690-20231212-182215-18c5d1cb-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)